gammapy/irf/background.py

```diff
@@ -114,3 +126,57 @@ def read(cls, filename, hdu='BACKGROUND'):
         filename = make_path(filename)
         hdulist = fits.open(str(filename))
         return cls.from_hdulist(hdulist, hdu=hdu)
+
+    def to_table(self, provenance=None):
+        """ Convert data to bintable HDU in CTA IRF format
+
+        Data format specification: :ref:`gadf:edisp_2d`
+
+        Parameters
+        ----------
+        provenance : `list`
+            dict containing required information for fits header
+
+        Examples
+        --------
+        Read energy dispersion IRF from disk:
+        from gammapy.irf import Background3D
+        
+        head = ([  
+            ('ORIGIN', 'IRAP', 'Name of organization making this file'),
+            ('DATE', '2017-09-27T12:02:24', 'File creation date (YYYY-MM-DDThh:mm:ss UTC)'),
+            ('TELESCOP', 'CTA', 'Name of telescope'),
+            ('INSTRUME', 'PROD3B', 'Name of instrument'),
+            ('DETNAM', 'NONE', 'Name of detector'),
+            ('HDUCLASS', 'OGIP', 'HDU class'),
+            ('HDUDOC', '???', 'HDU documentation'),
+            ('HDUCLAS1', 'RESPONSE', 'HDU class'),
+            ('HDUCLAS2', 'BKG', 'HDU class)
+            ...])
+            
+        bkg = Background3D(detx_lo, detx_hi, dety_lo, dety_hi, e_true_lo, e_true_hi, data)
+        hdu = bgk.to_table(head)
+        prim_hdu = fits.PrimaryHDU()
+        fits.HDUList([prim_hdu, hdu]).writeto('irffile.fits')
+        """
+        c1 = fits.Column(name='DETX_LO', array=np.asarray([self.detx.lo]),
+                         format='{}E'.format(self.detx.nbins), unit='{}'.format(self.detx.unit))
+        c2 = fits.Column(name='DETX_HI', array=np.asarray([self.detx.hi]),
+                         format='{}E'.format(self.detx.nbins), unit='{}'.format(self.detx.unit))
+        c3 = fits.Column(name='DETY_LO', array=np.asarray([self.dety.lo]),
+                         format='{}E'.format(self.dety.nbins), unit='{}'.format(self.detx.unit))
+        c4 = fits.Column(name='DETY_HI', array=np.asarray([self.dety.hi]),
+                         format='{}E'.format(self.dety.nbins), unit='{}'.format(self.detx.unit))
+        c5 = fits.Column(name='ENERGY_LO', array=np.asarray([self.energy.lo]),
+                         format='{}E'.format(self.energy.nbins), unit='{}'.format(self.energy.unit))
+        c6 = fits.Column(name='ENERGY_HI', array=np.asarray([self.energy.hi]),
+                         format='{}E'.format(self.energy.nbins), unit='{}'.format(self.energy.unit))
+        c7 = fits.Column(name='BGD', array=np.asarray([self.data.data]),
+                         format='{}E'.format(self.energy.nbins * self.detx.nbins * self.dety.nbins),
+                         dim='({},{},{})'.format(self.detx.nbins, self.dety.nbins, self.energy.nbins),
+                         unit='{}'.format(self.data.data.unit))
```

Comment:
This is an example of the code I'd like to avoid. `astropy.table.Table` computes the FITS header keywords for format and dim automatically. unfortunately, unit isn't handled automatically, but for now we can do it via a utility function (http://docs.gammapy.org/en/latest/api/gammapy.utils.fits.table_to_fits_table.html) and then later we can contribute this to the Table <-> BinTable converter in Astropy core.

Reply:
Table is indeed more simple, I think I had switched to BinTable just because I couldn't get Table to take the units for each column, but at least it goes from 
```python
c1 = fits.Column(name='ENERGY_LO', array=np.asarray([self.energy.lo]),
                 format='{}E'.format(self.energy.nbins), unit='{}'.format(self.energy.unit))
c2 = fits.Column(name='ENERGY_HI', array=np.asarray([self.energy.hi]),
                 format='{}E'.format(self.energy.nbins), unit='{}'.format(self.energy.unit))
c3 = fits.Column(name='OFFSET_LO', array=np.asarray([self.offset.lo]),
                 format='{}E'.format(self.offset.nbins), unit='{}'.format(self.offset.unit))
c4 = fits.Column(name='OFFSET_HI', array=np.asarray([self.offset.hi]),
                 format='{}E'.format(self.offset.nbins), unit='{}'.format(self.offset.unit))
c5 = fits.Column(name='EFFAREA', array=np.asarray([self.data.data.T]),
                 format='{}E'.format(self.energy.nbins * self.offset.nbins),
                 dim='({},{})'.format(self.energy.nbins, self.offset.nbins),
                 unit='{}'.format(self.data.data.unit))
```
to 

```python
table = Table([[self.energy.lo], [self.energy.hi], [self.offset.lo], [self.offset.hi], [self.data.data.T]],
              names=('ENERGY_LO', 'ENERGY_HI', 'OFFSET_LO', 'OFFSET_HI', 'EFFAREA'), meta=self.meta)
```

I believe table_to_fits_table already has the option to get the hdu name from the meta file, I suppose something similar to this could be done with the units for now?

Reply:
I think I would write it like this:
```
table = Table(meta=self.meta)
```
and then start filling one column at a time.
(Table is basically a dict of column name strings mapped to `Column` objects that hold the data, so filling column by column is efficient).
```
In [11]: table['ENERG_LO'] = aeff.data.axis('energy').lo
```
Assigning a Quantity object like that will create a column object with the unit filled from the quantity, which is what we want here:
```
In [12]: table.info()
<Table length=42>
  name    dtype  unit
-------- ------- ----
ENERG_LO float32  TeV

In [13]: table['ENERG_LO']
Out[13]: 
<Column name='ENERG_LO' dtype='float32' unit='TeV' length=42>
0.0125893
0.0158489
0.0199526
0.0251189
0.0316228
0.0398107
0.0501187
0.0630957
0.0794328
      0.1
 0.125893
 0.158489
 0.199526
 0.251189
 0.316228
 0.398107
 0.501187
      ...
  5.01187
  6.30957
  7.94328
     10.0
  12.5893
  15.8489
  19.9526
  25.1189
  31.6228
  39.8107
  50.1187
  63.0957
  79.4328
    100.0
  125.893
  158.489

In [14]: type(table['ENERG_LO'])
Out[14]: astropy.table.column.Column
```

So `to_table` would return a `Table`.
And then `read` and `write` would call `read` and `write` on that Table object, or if that looses the units or other metadata, we should go via [gammapy.utils.fits.table_to_fits_table](https://gammapy.readthedocs.io/en/latest/api/gammapy.utils.fits.table_to_fits_table.html) and [gammapy.utils.fits.fits_table_to_table](https://gammapy.readthedocs.io/en/latest/api/gammapy.utils.fits.fits_table_to_table.html) (and if that still doesn't work properly, fix up those utility functions, but only do that work of proper Table to BinTable conversion once in a central place in those helper functions. If it's unclear or you run into any issues (not unheard of with FITS), comment here and I'll take a closer look.